File: src/priv/bajdcc/LALR1/interpret/module/ModuleProc.java
package priv.bajdcc.LALR1.interpret.module;

import priv.bajdcc.LALR1.grammar.Grammar;
import priv.bajdcc.LALR1.grammar.runtime.*;
import priv.bajdcc.LALR1.grammar.runtime.RuntimeException;
import priv.bajdcc.LALR1.grammar.runtime.data.RuntimeArray;
import priv.bajdcc.LALR1.grammar.runtime.data.RuntimeFuncObject;
import priv.bajdcc.util.ResourceLoader;

import java.math.BigInteger;
import java.util.List;

/**
 * 【模块】进程模块
 *
 * @author bajdcc
 */
public class ModuleProc implements IInterpreterModule {

	private static ModuleProc instance = new ModuleProc();

	public static ModuleProc getInstance() {
		return instance;
	}

	private static final int LOCK_TIME = 20;
	private static final int PIPE_READ_TIME = 5;

	@Override
	public String getModuleName() {
		return "sys.proc";
	}

	@Override
	public RuntimeCodePage getCodePage() throws Exception {
		String base = ResourceLoader.load(getClass());

		Grammar grammar = new Grammar(base);
		RuntimeCodePage page = grammar.getCodePage();
		IRuntimeDebugInfo info = page.getInfo();
		buildMethod(info);
		buildPipeMethod(info);
		buildShareMethod(info);

		return page;
	}

	private void buildMethod(IRuntimeDebugInfo info) {
		info.addExternalFunc("g_create_process", new IRuntimeDebugExec() {
			@Override
			public String getDoc() {
				return "创建进程";
			}

			@Override
			public RuntimeObjectType[] getArgsType() {
				return new RuntimeObjectType[] { RuntimeObjectType.kFunc };
			}

			@Override
			public RuntimeObject ExternalProcCall(List<RuntimeObject> args,
			                                      IRuntimeStatus status) throws Exception {
				RuntimeFuncObject func = (RuntimeFuncObject) args.get(0).getObj();
				return new RuntimeObject(BigInteger.valueOf(status.createProcess(func)));
			}
		});
		info.addExternalFunc("g_create_process_args", new IRuntimeDebugExec() {
			@Override
			public String getDoc() {
				return "创建进程带参数";
			}

			@Override
			public RuntimeObjectType[] getArgsType() {
				return new RuntimeObjectType[] { RuntimeObjectType.kFunc, RuntimeObjectType.kObject };
			}

			@Override
			public RuntimeObject ExternalProcCall(List<RuntimeObject> args,
			                                      IRuntimeStatus status) throws Exception {
				RuntimeFuncObject func = (RuntimeFuncObject) args.get(0).getObj();
				RuntimeObject obj = args.get(1);
				return new RuntimeObject(BigInteger.valueOf(status.createProcess(func, obj)));
			}
		});
		info.addExternalFunc("g_create_user_process", new IRuntimeDebugExec() {
			@Override
			public String getDoc() {
				return "创建用户态进程";
			}

			@Override
			public RuntimeObjectType[] getArgsType() {
				return new RuntimeObjectType[] { RuntimeObjectType.kFunc };
			}

			@Override
			public RuntimeObject ExternalProcCall(List<RuntimeObject> args,
			                                      IRuntimeStatus status) throws Exception {
				RuntimeFuncObject func = (RuntimeFuncObject) args.get(0).getObj();
				return new RuntimeObject(BigInteger.valueOf(status.createUsrProcess(func)));
			}
		});
		info.addExternalFunc("g_create_user_process_args", new IRuntimeDebugExec() {
			@Override
			public String getDoc() {
				return "创建用户态进程带参数";
			}

			@Override
			public RuntimeObjectType[] getArgsType() {
				return new RuntimeObjectType[] { RuntimeObjectType.kFunc, RuntimeObjectType.kObject };
			}

			@Override
			public RuntimeObject ExternalProcCall(List<RuntimeObject> args,
			                                      IRuntimeStatus status) throws Exception {
				RuntimeFuncObject func = (RuntimeFuncObject) args.get(0).getObj();
				RuntimeObject obj = args.get(1);
				return new RuntimeObject(BigInteger.valueOf(status.createUsrProcess(func, obj)));
			}
		});
		info.addExternalFunc("g_get_user_procs", new IRuntimeDebugExec() {
			@Override
			public String getDoc() {
				return "获取用户态进程ID列表";
			}

			@Override
			public RuntimeObjectType[] getArgsType() {
				return null;
			}

			@Override
			public RuntimeObject ExternalProcCall(List<RuntimeObject> args,
			                                      IRuntimeStatus status) throws Exception {
				RuntimeArray arr = new RuntimeArray();
				List<Integer> list = status.getUsrProcs();
				for (Integer pid : list) {
					arr.add(new RuntimeObject(pid));
				}
				return new RuntimeObject(arr);
			}
		});
		info.addExternalFunc("g_run_user", new IRuntimeDebugExec() {
			@Override
			public String getDoc() {
				return "运行用户态进程";
			}

			@Override
			public RuntimeObjectType[] getArgsType() {
				return new RuntimeObjectType[] { RuntimeObjectType.kPtr };
			}

			@Override
			public RuntimeObject ExternalProcCall(List<RuntimeObject> args,
			                                      IRuntimeStatus status) throws Exception {
				int pid = (int) args.get(0).getObj();
				return new RuntimeObject(BigInteger.valueOf(status.stepUsrProcess(pid)));
			}
		});
		info.addExternalFunc("g_get_pid", new IRuntimeDebugExec() {
			@Override
			public String getDoc() {
				return "获取进程ID";
			}

			@Override
			public RuntimeObjectType[] getArgsType() {
				return null;
			}

			@Override
			public RuntimeObject ExternalProcCall(List<RuntimeObject> args,
			                                      IRuntimeStatus status) throws Exception {
				return new RuntimeObject(BigInteger.valueOf(status.getPid()));
			}
		});
		info.addExternalFunc("g_get_parent_pid", new IRuntimeDebugExec() {
			@Override
			public String getDoc() {
				return "获取父进程ID";
			}

			@Override
			public RuntimeObjectType[] getArgsType() {
				return null;
			}

			@Override
			public RuntimeObject ExternalProcCall(List<RuntimeObject> args,
			                                      IRuntimeStatus status) throws Exception {
				return new RuntimeObject(BigInteger.valueOf(status.getParentPid()));
			}
		});
		info.addExternalFunc("g_get_process_priority", new IRuntimeDebugExec() {
			@Override
			public String getDoc() {
				return "获取进程优先级";
			}

			@Override
			public RuntimeObjectType[] getArgsType() {
				return null;
			}

			@Override
			public RuntimeObject ExternalProcCall(List<RuntimeObject> args,
			                                      IRuntimeStatus status) throws Exception {
				return new RuntimeObject(BigInteger.valueOf(status.getPriority()));
			}
		});
		info.addExternalFunc("g_set_process_priority", new IRuntimeDebugExec() {
			@Override
			public String getDoc() {
				return "设置进程优先级";
			}

			@Override
			public RuntimeObjectType[] getArgsType() {
				return new RuntimeObjectType[] { RuntimeObjectType.kInt };
			}

			@Override
			public RuntimeObject ExternalProcCall(List<RuntimeObject> args,
			                                      IRuntimeStatus status) throws Exception {
				BigInteger priority = (BigInteger) args.get(0).getObj();
				return new RuntimeObject(status.setPriority(priority.intValue()));
			}
		});
		info.addExternalFunc("g_join_process_once", new IRuntimeDebugExec() {
			@Override
			public String getDoc() {
				return "进程等待";
			}

			@Override
			public RuntimeObjectType[] getArgsType() {
				return new RuntimeObjectType[] { RuntimeObjectType.kInt };
			}

			@Override
			public RuntimeObject ExternalProcCall(List<RuntimeObject> args,
			                                      IRuntimeStatus status) throws Exception {
				BigInteger pid = (BigInteger) args.get(0).getObj();
				return new RuntimeObject(status.getService().getProcessService().join(pid.intValue(), status.getPid()));
			}
		});
		info.addExternalFunc("g_live_process", new IRuntimeDebugExec() {
			@Override
			public String getDoc() {
				return "进程存活";
			}

			@Override
			public RuntimeObjectType[] getArgsType() {
				return new RuntimeObjectType[]{RuntimeObjectType.kInt};
			}

			@Override
			public RuntimeObject ExternalProcCall(List<RuntimeObject> args,
			                                      IRuntimeStatus status) throws Exception {
				BigInteger pid = (BigInteger) args.get(0).getObj();
				return new RuntimeObject(status.getService().getProcessService().live(pid.intValue()));
			}
		});
		info.addExternalFunc("g_sleep", new IRuntimeDebugExec() {
			@Override
			public String getDoc() {
				return "进程睡眠";
			}

			@Override
			public RuntimeObjectType[] getArgsType() {
				return new RuntimeObjectType[] { RuntimeObjectType.kInt };
			}

			@Override
			public RuntimeObject ExternalProcCall(List<RuntimeObject> args,
			                                      IRuntimeStatus status) throws Exception {
				BigInteger turn = (BigInteger) args.get(0).getObj();
				int time = turn.intValue();
				return new RuntimeObject(BigInteger.valueOf(
						status.getService().getProcessService().sleep(status.getPid(), time > 0 ? time : 0)));
			}
		});
		info.addExternalFunc("g_query_usr_proc", new IRuntimeDebugExec() {
			@Override
			public String getDoc() {
				return "枚举用户态进程";
			}

			@Override
			public RuntimeObjectType[] getArgsType() {
				return null;
			}

			@Override
			public RuntimeObject ExternalProcCall(List<RuntimeObject> args,
			                                      IRuntimeStatus status) throws Exception {
				return new RuntimeObject(getProcInfo(status, status.getUsrProcs()));
			}
		});
		info.addExternalFunc("g_query_sys_proc", new IRuntimeDebugExec() {
			@Override
			public String getDoc() {
				return "枚举内核态进程";
			}

			@Override
			public RuntimeObjectType[] getArgsType() {
				return null;
			}

			@Override
			public RuntimeObject ExternalProcCall(List<RuntimeObject> args,
			                                      IRuntimeStatus status) throws Exception {
				return new RuntimeObject(getProcInfo(status, status.getSysProcs()));
			}
		});
		info.addExternalFunc("g_set_process_desc", new IRuntimeDebugExec() {
			@Override
			public String getDoc() {
				return "设置进程说明";
			}

			@Override
			public RuntimeObjectType[] getArgsType() {
				return new RuntimeObjectType[]{RuntimeObjectType.kString};
			}

			@Override
			public RuntimeObject ExternalProcCall(List<RuntimeObject> args,
			                                      IRuntimeStatus status) throws Exception {
				status.setProcDesc(String.valueOf(args.get(0).getObj()));
				return null;
			}
		});
	}

	private void buildPipeMethod(IRuntimeDebugInfo info) {
		info.addExternalFunc("g_create_pipe", new IRuntimeDebugExec() {
			@Override
			public String getDoc() {
				return "创建管道";
			}

			@Override
			public RuntimeObjectType[] getArgsType() {
				return new RuntimeObjectType[] { RuntimeObjectType.kString };
			}

			@Override
			public RuntimeObject ExternalProcCall(List<RuntimeObject> args,
												  IRuntimeStatus status) throws Exception {
				String name = args.get(0).getObj().toString();
				int handle = status.getService().getPipeService().create(name);
				if (handle == -1)
					status.err(RuntimeException.RuntimeError.MAX_HANDLE);
				return new RuntimeObject(handle);
			}
		});
		info.addExternalFunc("g_query_pipe", new IRuntimeDebugExec() {
			@Override
			public String getDoc() {
				return "查询管道";
			}

			@Override
			public RuntimeObjectType[] getArgsType() {
				return new RuntimeObjectType[] { RuntimeObjectType.kString };
			}

			@Override
			public RuntimeObject ExternalProcCall(List<RuntimeObject> args,
												  IRuntimeStatus status) throws Exception {
				String name = args.get(0).getObj().toString();
				return new RuntimeObject(status.getService().getPipeService().query(name));
			}
		});
		info.addExternalFunc("g_destroy_pipe_once", new IRuntimeDebugExec() {
			@Override
			public String getDoc() {
				return "销毁管道";
			}

			@Override
			public RuntimeObjectType[] getArgsType() {
				return new RuntimeObjectType[] { RuntimeObjectType.kPtr };
			}

			@Override
			public RuntimeObject ExternalProcCall(List<RuntimeObject> args,
												  IRuntimeStatus status) throws Exception {
				int handle = (int) args.get(0).getObj();
				return new RuntimeObject(status.getService().getPipeService().destroy(handle));
			}
		});
		info.addExternalFunc("g_destroy_pipe_by_name_once", new IRuntimeDebugExec() {
			@Override
			public String getDoc() {
				return "销毁管道";
			}

			@Override
			public RuntimeObjectType[] getArgsType() {
				return new RuntimeObjectType[]{RuntimeObjectType.kString};
			}

			@Override
			public RuntimeObject ExternalProcCall(List<RuntimeObject> args,
			                                      IRuntimeStatus status) throws Exception {
				String name = String.valueOf(args.get(0).getObj());
				return new RuntimeObject(status.getService().getPipeService().destroyByName(status.getPid(), name));
			}
		});
		info.addExternalFunc("g_wait_pipe_empty", new IRuntimeDebugExec() {
			@Override
			public String getDoc() {
				return "等待管道为空";
			}

			@Override
			public RuntimeObjectType[] getArgsType() {
				return new RuntimeObjectType[] { RuntimeObjectType.kPtr };
			}

			@Override
			public RuntimeObject ExternalProcCall(List<RuntimeObject> args,
												  IRuntimeStatus status) throws Exception {
				int handle = (int) args.get(0).getObj();
				status.getService().getProcessService().sleep(status.getPid(), PIPE_READ_TIME);
				return new RuntimeObject(!status.getService().getPipeService().isEmpty(handle));
			}
		});
		info.addExternalFunc("g_read_pipe_char", new IRuntimeDebugExec() {
			@Override
			public String getDoc() {
				return "管道读";
			}

			@Override
			public RuntimeObjectType[] getArgsType() {
				return new RuntimeObjectType[] { RuntimeObjectType.kPtr };
			}

			@Override
			public RuntimeObject ExternalProcCall(List<RuntimeObject> args,
												  IRuntimeStatus status) throws Exception {
				int handle = (int) args.get(0).getObj();
				char ch = status.getService().getPipeService().read(status.getPid(), handle);
				return new RuntimeObject(ch);
			}
		});
		info.addExternalFunc("g_write_pipe_char", new IRuntimeDebugExec() {
			@Override
			public String getDoc() {
				return "管道写";
			}

			@Override
			public RuntimeObjectType[] getArgsType() {
				return new RuntimeObjectType[] { RuntimeObjectType.kPtr, RuntimeObjectType.kChar };
			}

			@Override
			public RuntimeObject ExternalProcCall(List<RuntimeObject> args,
												  IRuntimeStatus status) throws Exception {
				int handle = (int) args.get(0).getObj();
				char ch = (char) args.get(1).getObj();
				return new RuntimeObject(status.getService().getPipeService().write(status.getPid(), handle, ch));
			}
		});
	}

	private void buildShareMethod(IRuntimeDebugInfo info) {
		info.addExternalFunc("g_start_share", new IRuntimeDebugExec() {
			@Override
			public String getDoc() {
				return "创建共享";
			}

			@Override
			public RuntimeObjectType[] getArgsType() {
				return new RuntimeObjectType[] { RuntimeObjectType.kString, RuntimeObjectType.kObject };
			}

			@Override
			public RuntimeObject ExternalProcCall(List<RuntimeObject> args,
			                                      IRuntimeStatus status) throws Exception {
				String name = args.get(0).getObj().toString();
				int result = status.getService().getShareService().startSharing(name, args.get(1));
				if (result == -1)
					status.err(RuntimeException.RuntimeError.MAX_HANDLE);
				if (result == 0)
					status.err(RuntimeException.RuntimeError.DUP_SHARE_NAME);
				return new RuntimeObject(BigInteger.valueOf(result));
			}
		});
		info.addExternalFunc("g_create_share", new IRuntimeDebugExec() {
			@Override
			public String getDoc() {
				return "创建共享";
			}

			@Override
			public RuntimeObjectType[] getArgsType() {
				return new RuntimeObjectType[]{RuntimeObjectType.kString, RuntimeObjectType.kObject};
			}

			@Override
			public RuntimeObject ExternalProcCall(List<RuntimeObject> args,
			                                      IRuntimeStatus status) throws Exception {
				String name = args.get(0).getObj().toString();
				int result = status.getService().getShareService().createSharing(name, args.get(1));
				if (result == -1)
					status.err(RuntimeException.RuntimeError.MAX_HANDLE);
				return new RuntimeObject(BigInteger.valueOf(result));
			}
		});
		info.addExternalFunc("g_query_share", new IRuntimeDebugExec() {
			@Override
			public String getDoc() {
				return "查询共享";
			}

			@Override
			public RuntimeObjectType[] getArgsType() {
				return new RuntimeObjectType[] { RuntimeObjectType.kString };
			}

			@Override
			public RuntimeObject ExternalProcCall(List<RuntimeObject> args,
			                                      IRuntimeStatus status) throws Exception {
				String name = args.get(0).getObj().toString();
				return status.getService().getShareService().getSharing(name, false);
			}
		});
		info.addExternalFunc("g_reference_share", new IRuntimeDebugExec() {
			@Override
			public String getDoc() {
				return "引用共享";
			}

			@Override
			public RuntimeObjectType[] getArgsType() {
				return new RuntimeObjectType[] { RuntimeObjectType.kString };
			}

			@Override
			public RuntimeObject ExternalProcCall(List<RuntimeObject> args,
			                                      IRuntimeStatus status) throws Exception {
				String name = args.get(0).getObj().toString();
				return status.getService().getShareService().getSharing(name, true);
			}
		});
		info.addExternalFunc("g_stop_share", new IRuntimeDebugExec() {
			@Override
			public String getDoc() {
				return "停止共享";
			}

			@Override
			public RuntimeObjectType[] getArgsType() {
				return new RuntimeObjectType[] { RuntimeObjectType.kString };
			}

			@Override
			public RuntimeObject ExternalProcCall(List<RuntimeObject> args,
			                                      IRuntimeStatus status) throws Exception {
				String name = args.get(0).getObj().toString();
				int result = status.getService().getShareService().stopSharing(name);
				if (result == -1)
					status.err(RuntimeException.RuntimeError.INVALID_SHARE_NAME);
				if (result == 2)
					status.err(RuntimeException.RuntimeError.INVALID_REFERENCE);
				return new RuntimeObject(result == 1);
			}
		});
		info.addExternalFunc("g_try_lock_share", new IRuntimeDebugExec() {
			@Override
			public String getDoc() {
				return "尝试锁定共享变量";
			}

			@Override
			public RuntimeObjectType[] getArgsType() {
				return new RuntimeObjectType[] { RuntimeObjectType.kString };
			}

			@Override
			public RuntimeObject ExternalProcCall(List<RuntimeObject> args,
			                                      IRuntimeStatus status) throws Exception {
				String name = args.get(0).getObj().toString();
				if (status.getService().getShareService().isLocked(name)) {
					status.getService().getProcessService().sleep(status.getPid(), LOCK_TIME);
					return new RuntimeObject(true);
				}
				status.getService().getShareService().setLocked(name, true);
				return new RuntimeObject(false);
			}
		});
		info.addExternalFunc("g_unlock_share", new IRuntimeDebugExec() {
			@Override
			public String getDoc() {
				return "解锁共享变量";
			}

			@Override
			public RuntimeObjectType[] getArgsType() {
				return new RuntimeObjectType[] { RuntimeObjectType.kString };
			}

			@Override
			public RuntimeObject ExternalProcCall(List<RuntimeObject> args,
			                                      IRuntimeStatus status) throws Exception {
				String name = args.get(0).getObj().toString();
				if (status.getService().getShareService().isLocked(name)) {
					status.getService().getShareService().setLocked(name, false);
					return new RuntimeObject(true);
				}
				return new RuntimeObject(false);
			}
		});
	}

	private static RuntimeArray getProcInfo(IRuntimeStatus status, List<Integer> pids) {
		RuntimeArray array = new RuntimeArray();
		array.add(new RuntimeObject(String.format(" %s  %-5s   %-15s   %-20s   %s",
				"#", "Pid", "Name", "Function", "Description")));
		for (int pid : pids) {
			Object[] objs = status.getProcInfoById(pid);
			array.add(new RuntimeObject(String.format(" %s  %-5d   %-15s   %-20s   %s",
					objs[0], pid, objs[1], objs[2], objs[3])));
		}
		return array;
	}
}

File: src/main/java/com/bajdcc/LALR1/interpret/module/ModuleProc.kt
package com.bajdcc.LALR1.interpret.module

import com.bajdcc.LALR1.grammar.Grammar
import com.bajdcc.LALR1.grammar.runtime.*
import com.bajdcc.LALR1.grammar.runtime.data.RuntimeArray
import com.bajdcc.LALR1.grammar.runtime.data.RuntimeFuncObject
import com.bajdcc.util.ResourceLoader

@Suppress("UNUSED_ANONYMOUS_PARAMETER")
/**
 * 【模块】进程模块
 *
 * @author bajdcc
 */
class ModuleProc : IInterpreterModule {
    private var runtimeCodePage: RuntimeCodePage? = null

    override val moduleName: String
        get() = "sys.proc"

    override val moduleCode: String
        get() = ResourceLoader.load(javaClass)

    override val codePage: RuntimeCodePage
        @Throws(Exception::class)
        get() {
            if (runtimeCodePage != null)
                return runtimeCodePage!!

            val base = ResourceLoader.load(javaClass)

            val grammar = Grammar(base)
            val page = grammar.codePage
            val info = page.info
            buildMethod(info)
            buildPipeMethod(info)
            buildShareMethod(info)

            runtimeCodePage = page
            return page
        }

    private fun buildMethod(info: IRuntimeDebugInfo) {
        info.addExternalFunc("g_create_process",
                RuntimeDebugExec("创建进程", arrayOf(RuntimeObjectType.kFunc))
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        val func = args[0].obj as RuntimeFuncObject?
                        RuntimeObject(status.createProcess(func!!).toLong())
                    }
                })
        info.addExternalFunc("g_create_process_args",
                RuntimeDebugExec("创建进程带参数", arrayOf(RuntimeObjectType.kFunc, RuntimeObjectType.kObject))
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        val func = args[0].obj as RuntimeFuncObject?
                        val obj = args[1]
                        RuntimeObject(status.createProcess(func!!, obj).toLong())
                    }
                })
        info.addExternalFunc("g_create_user_process",
                RuntimeDebugExec("创建用户态进程", arrayOf(RuntimeObjectType.kFunc))
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        val func = args[0].obj as RuntimeFuncObject?
                        RuntimeObject(status.createUsrProcess(func!!).toLong())
                    }
                })
        info.addExternalFunc("g_create_user_process_args",
                RuntimeDebugExec("创建用户态进程带参数", arrayOf(RuntimeObjectType.kFunc, RuntimeObjectType.kObject))
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        val func = args[0].obj as RuntimeFuncObject?
                        val obj = args[1]
                        RuntimeObject(status.createUsrProcess(func!!, obj).toLong())
                    }
                })
        info.addExternalFunc("g_get_user_procs",
                RuntimeDebugExec("获取用户态进程ID列表")
                { args: List<RuntimeObject>, status: IRuntimeStatus -> RuntimeObject(status.usrProcs.map { RuntimeObject(it) }) })
        info.addExternalFunc("g_get_pid",
                RuntimeDebugExec("获取进程ID")
                { args: List<RuntimeObject>, status: IRuntimeStatus -> RuntimeObject(status.pid.toLong()) })
        info.addExternalFunc("g_get_parent_pid",
                RuntimeDebugExec("获取父进程ID")
                { args: List<RuntimeObject>, status: IRuntimeStatus -> RuntimeObject(status.parentPid.toLong()) })
        info.addExternalFunc("g_get_process_priority",
                RuntimeDebugExec("获取进程优先级")
                { args: List<RuntimeObject>, status: IRuntimeStatus -> RuntimeObject(status.priority.toLong()) })
        info.addExternalFunc("g_set_process_priority",
                RuntimeDebugExec("设置进程优先级", arrayOf(RuntimeObjectType.kInt))
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        val priority = args[0].long
                        RuntimeObject(status.setPriority(priority.toInt()))
                    }
                })
        info.addExternalFunc("g_join_process_once",
                RuntimeDebugExec("进程等待", arrayOf(RuntimeObjectType.kInt))
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        val pid = args[0].long
                        RuntimeObject(status.service.processService.join(pid.toInt(), status.pid))
                    }
                })
        info.addExternalFunc("g_live_process",
                RuntimeDebugExec("进程存活", arrayOf(RuntimeObjectType.kInt))
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        val pid = args[0].long
                        RuntimeObject(status.service.processService.live(pid.toInt()))
                    }
                })
        info.addExternalFunc("g_sleep",
                RuntimeDebugExec("进程睡眠", arrayOf(RuntimeObjectType.kInt))
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        val time = args[0].int
                        RuntimeObject(status.service.processService.sleep(status.pid, if (time > 0) time else 0).toLong())
                    }
                })
        info.addExternalFunc("g_query_usr_proc",
                RuntimeDebugExec("枚举用户态进程")
                { args: List<RuntimeObject>, status: IRuntimeStatus -> RuntimeObject(ProcInfoHelper.getProcInfo(status, status.usrProcs)) })
        info.addExternalFunc("g_query_sys_proc",
                RuntimeDebugExec("枚举内核态进程")
                { args: List<RuntimeObject>, status: IRuntimeStatus -> RuntimeObject(ProcInfoHelper.getProcInfo(status, status.sysProcs)) })
        info.addExternalFunc("g_query_all_proc",
                RuntimeDebugExec("枚举进程")
                { args: List<RuntimeObject>, status: IRuntimeStatus -> RuntimeObject(ProcInfoHelper.getProcInfoAll(status)) })
        info.addExternalFunc("g_set_process_desc",
                RuntimeDebugExec("设置进程说明", arrayOf(RuntimeObjectType.kString))
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        status.setProcDesc(args[0].obj.toString())
                        null
                    }
                })
    }

    private fun buildPipeMethod(info: IRuntimeDebugInfo) {
        info.addExternalFunc("g_create_pipe",
                RuntimeDebugExec("创建管道", arrayOf(RuntimeObjectType.kString))
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        val name = args[0].string
                        val handle = status.service.pipeService.create(name, status.page)
                        if (handle == -1)
                            status.err(RuntimeException.RuntimeError.MAX_HANDLE)
                        RuntimeObject(handle)
                    }
                })
        info.addExternalFunc("g_query_pipe",
                RuntimeDebugExec("查询管道", arrayOf(RuntimeObjectType.kString))
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        val name = args[0].string
                        RuntimeObject(status.service.pipeService.query(name))
                    }
                })
        info.addExternalFunc("g_destroy_pipe_once",
                RuntimeDebugExec("销毁管道", arrayOf(RuntimeObjectType.kPtr))
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        val handle = args[0].int
                        RuntimeObject(status.service.pipeService.destroy(handle))
                    }
                })
        info.addExternalFunc("g_destroy_pipe_by_name_once",
                RuntimeDebugExec("销毁管道", arrayOf(RuntimeObjectType.kString))
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        val name = args[0].obj.toString()
                        RuntimeObject(status.service.pipeService.destroyByName(status.pid, name))
                    }
                })
        info.addExternalFunc("g_wait_pipe_empty",
                RuntimeDebugExec("等待管道为空", arrayOf(RuntimeObjectType.kPtr))
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        val handle = args[0].int
                        status.service.processService.sleep(status.pid, PIPE_READ_TIME)
                        RuntimeObject(status.service.pipeService.hasData(handle))
                    }
                })
        info.addExternalFunc("g_read_pipe_char",
                RuntimeDebugExec("管道读", arrayOf(RuntimeObjectType.kPtr))
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        val handle = args[0].int
                        val ch = status.service.pipeService.read(status.pid, handle)
                        RuntimeObject(ch)
                    }
                })
        info.addExternalFunc("g_read_pipe_char_no_block",
                RuntimeDebugExec("管道读（非阻塞）", arrayOf(RuntimeObjectType.kPtr))
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        val handle = args[0].int
                        val ch = status.service.pipeService.readNoBlock(status.pid, handle)
                        RuntimeObject(ch)
                    }
                })
        info.addExternalFunc("g_write_pipe_char",
                RuntimeDebugExec("管道写", arrayOf(RuntimeObjectType.kPtr, RuntimeObjectType.kChar))
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        val handle = args[0].int
                        val ch = args[1].char
                        RuntimeObject(status.service.pipeService.write(handle, ch))
                    }
                })
    }

    private fun buildShareMethod(info: IRuntimeDebugInfo) {
        info.addExternalFunc("g_start_share",
                RuntimeDebugExec("创建共享", arrayOf(RuntimeObjectType.kString, RuntimeObjectType.kObject))
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        val name = args[0].string
                        val result = status.service.shareService.startSharing(name, args[1], status.page)
                        if (result == -1)
                            status.err(RuntimeException.RuntimeError.MAX_HANDLE, name)
                        if (result == 0)
                            status.err(RuntimeException.RuntimeError.DUP_SHARE_NAME, name)
                        RuntimeObject(result.toLong())
                    }
                })
        info.addExternalFunc("g_create_share",
                RuntimeDebugExec("创建共享", arrayOf(RuntimeObjectType.kString, RuntimeObjectType.kObject))
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        val name = args[0].string
                        val result = status.service.shareService.createSharing(name, args[1], status.page)
                        if (result == -1)
                            status.err(RuntimeException.RuntimeError.MAX_HANDLE, name)
                        RuntimeObject(result.toLong())
                    }
                })
        info.addExternalFunc("g_query_share",
                RuntimeDebugExec("查询共享", arrayOf(RuntimeObjectType.kString))
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        val name = args[0].string
                        status.service.shareService.getSharing(name, false)
                    }
                })
        info.addExternalFunc("g_reference_share",
                RuntimeDebugExec("引用共享", arrayOf(RuntimeObjectType.kString))
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        val name = args[0].string
                        status.service.shareService.getSharing(name, true)
                    }
                })
        info.addExternalFunc("g_stop_share",
                RuntimeDebugExec("停止共享", arrayOf(RuntimeObjectType.kString))
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        val name = args[0].string
                        val result = status.service.shareService.stopSharing(name)
                        if (result == -1)
                            status.err(RuntimeException.RuntimeError.INVALID_SHARE_NAME, name)
                        if (result == 2)
                            status.err(RuntimeException.RuntimeError.INVALID_REFERENCE, name)
                        RuntimeObject(result == 1)
                    }
                })
        info.addExternalFunc("g_try_lock_share",
                RuntimeDebugExec("尝试锁定共享变量", arrayOf(RuntimeObjectType.kString))
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        val name = args[0].string
                        if (status.service.shareService.isLocked(name)) {
                            status.service.processService.sleep(status.pid, LOCK_TIME)
                            RuntimeObject(true)
                        } else {
                            status.service.shareService.setLocked(name, true)
                            RuntimeObject(false)
                        }
                    }
                })
        info.addExternalFunc("g_unlock_share",
                RuntimeDebugExec("解锁共享变量", arrayOf(RuntimeObjectType.kString))
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        val name = args[0].string
                        if (status.service.shareService.isLocked(name)) {
                            status.service.shareService.setLocked(name, false)
                            RuntimeObject(true)
                        } else {
                            RuntimeObject(false)
                        }
                    }
                })
        info.addExternalFunc("g_proc_exec",
                RuntimeDebugExec("运行用户态代码", arrayOf(RuntimeObjectType.kString))
                { args: List<RuntimeObject>, status: IRuntimeStatus -> RuntimeObject(status.ring3.exec(args[0].string).toLong()) })
        info.addExternalFunc("g_proc_exec_file",
                RuntimeDebugExec("运行用户态代码", arrayOf(RuntimeObjectType.kString, RuntimeObjectType.kString))
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    RuntimeObject(status.ring3.execFile(
                            args[0].obj.toString(),
                            args[1].obj.toString()).toLong())
                })
        info.addExternalFunc("g_proc_kill",
                RuntimeDebugExec("强制结束用户态进程", arrayOf(RuntimeObjectType.kInt))
                { args: List<RuntimeObject>, status: IRuntimeStatus ->
                    run {
                        val pid = args[0].long
                        RuntimeObject(status.service.processService.ring3Kill(pid.toInt(), "强制退出").toLong())
                    }
                })
    }

    internal object ProcInfoHelper {
        fun getProcInfo(status: IRuntimeStatus, pids: List<Int>): RuntimeArray {
            return getProcInfo2(getProcInfo3(status, pids))
        }

        fun getProcInfoAll(status: IRuntimeStatus): RuntimeArray {
            return getProcInfo2(getProcInfo4(status))
        }

        fun getProcInfo2(objs: List<Array<Any>>): RuntimeArray {
            val array = RuntimeArray()
            array.add(RuntimeObject(String.format(" %s  %s %-5s   %-15s   %-25s   %s",
                    " ", "环", "标识", "名称", "过程", "描述")))
            for (obj in objs) {
                var name = obj[4].toString()
                name = name.substring(0, Math.min(name.length, 20))
                array.add(RuntimeObject(String.format(" %s  %s %5s   %-15s   %-25s   %s",
                        obj[0], obj[1], obj[2], obj[3], name, obj[5])))
            }
            return array
        }

        fun getProcInfo3(status: IRuntimeStatus, pids: List<Int>): List<Array<Any>> {
            val objs = arrayListOf<Array<Any>>()
            for (pid in pids) {
                objs.add(status.getProcInfoById(pid))
            }
            return objs
        }

        fun getProcInfo4(status: IRuntimeStatus): List<Array<Any>> {
            return status.service.processService.procInfoCache
        }
    }

    companion object {

        val instance = ModuleProc()

        private const val LOCK_TIME = 20
        private const val PIPE_READ_TIME = 5
    }
}
